portfolio/src/main/java/com/google/sps/data/CalendarEventsGroup.java

```diff
@@ -0,0 +1,102 @@
+public class CalendarEventsGroup {
+
+  // This variable that represents a group of calendar events will
+  // always be sorted in ascending order based on their start time.
+  public List<CalendarEvent> events;
+
+  // These two variables represent the start and end of
+  // possible schedule time. When these times are passed in
+  // to instantiate a this class, these times are
+  // already adjusted based on user's working hour times
+  // and the current timestamp of when the user starts
+  // the scheduling process.
+  public final Instant overallStartTime;
+  public final Instant overallEndTime;
+
+  // Comparator for sorting events by start time
+  private static final Comparator<CalendarEvent> sortByEventStartTimeAscending =
+      Comparator.comparing(CalendarEvent::getStartTime);
+
+  /**
+   * @param events: a list of CalendarEvents' that represents the events already scheduled for the
+   *     user.
+   * @param startTime: start of possible scheduling blocks, of type Instant. This is the earlier of
+   *     working hour start time and the current timestamp.
+   * @param endTime: end of possible scheduling blocks, of type Instant.
+   */
+  public CalendarEventsGroup(
+      List<CalendarEvent> events, Instant overallStartTime, Instant overallEndTime) {
+    if (events == null) {
+      throw new IllegalArgumentException("Events cannot be null");
+    }
+    if (overallStartTime == null) {
+      throw new IllegalArgumentException("Overall start of scheduling hours cannot be null");
+    }
+    if (overallEndTime == null) {
+      throw new IllegalArgumentException("Overall end of scheduling hours cannot be null");
+    }
+    if (overallStartTime.isAfter(overallEndTime)) {
+      throw new IllegalArgumentException("Start time cannot be after end time");
+    }
+    this.events = events;
+    this.overallStartTime = overallStartTime;
+    this.overallEndTime = overallEndTime;
+  }
+
+  /**
+   * Calculates a TimeRange list which represents the periods of time that are empty of events and
+   * lie completely inside the possible scheduling hours. Originally written by tomasalvarez,
+   * modified by hollyyuqizheng.
+   */
+  public List<TimeRange> getFreeTimeRanges() {
+    Collections.sort(events, sortByEventStartTimeAscending);
+
+    List<TimeRange> possibleTimes = new ArrayList<TimeRange>();
+
+    // This represents the earliest time that we can schedule a window for the
+    // meeting. As events are processed, this changes to their end times.
+    Instant earliestNonScheduledInstant = overallStartTime;
+
+    for (CalendarEvent event : events) {
+      // Make sure that there is some time between the events and it is not
+      // later than the person's scheduling hours' ending time.
+      if (event.getStartTime().isAfter(earliestNonScheduledInstant)) {
+        if (event.getEndTime().isAfter(overallEndTime)) {
+          possibleTimes.add(TimeRange.fromStartEnd(earliestNonScheduledInstant, overallEndTime));
```

Comment:
This line doesn't seem correct to me. For example, suppose:
earliestNonScheduledInstant is t1
event.getStartTime() is t2
overallEndTime is t3
event.getEndTime() is t4

Here you're considering (t1, t3) as free. But it isn't, because the event starts at t2.

Reply:
I see. I updated the if-statement and added new tests.